src/cascadia/TerminalSettingsEditor/KeyChordListener.cpp

```diff
@@ -0,0 +1,143 @@
+                {
+                    flags |= VirtualKeyModifiers::Control;
+                    break;
+                }
+                case VirtualKey::Menu:
+                case VirtualKey::LeftMenu:
+                case VirtualKey::RightMenu:
+                {
+                    flags |= VirtualKeyModifiers::Menu;
+                    break;
+                }
+                case VirtualKey::Shift:
+                case VirtualKey::LeftShift:
+                {
+                    flags |= VirtualKeyModifiers::Shift;
+                    break;
+                }
+                case VirtualKey::LeftWindows:
+                case VirtualKey::RightWindows:
+                {
+                    flags |= VirtualKeyModifiers::Windows;
+                    break;
+                }
+                }
+            }
+        }
+        return flags;
+    }
+
+    KeyChordListener::KeyChordListener()
+    {
+        InitializeComponent();
+        _InitializeProperties();
+    }
+
+    void KeyChordListener::_InitializeProperties()
+    {
+        // Initialize any KeyChordListener dependency properties here.
+        // This performs a lazy load on these properties, instead of
+        // initializing them when the DLL loads.
+        if (!_KeysProperty)
+        {
+            _KeysProperty =
+                DependencyProperty::Register(
+                    L"Keys",
+                    xaml_typename<Control::KeyChord>(),
+                    xaml_typename<Editor::KeyChordListener>(),
+                    PropertyMetadata{ nullptr, PropertyChangedCallback{ &KeyChordListener::_OnKeysChanged } });
+        }
+    }
+
+    void KeyChordListener::_OnKeysChanged(DependencyObject const& d, DependencyPropertyChangedEventArgs const& e)
+    {
+        if (auto control{ d.try_as<Editor::KeyChordListener>() })
+        {
+            auto controlImpl{ get_self<KeyChordListener>(control) };
+            TextBox tb{ controlImpl->FindName(L"KeyChordTextBox").as<TextBox>() };
+            tb.Text(Model::KeyChordSerialization::ToString(unbox_value<Control::KeyChord>(e.NewValue())));
+            if (auto automationPeer{ Automation::Peers::FrameworkElementAutomationPeer::FromElement(tb) })
+            {
+                automationPeer.RaiseNotificationEvent(
+                    Automation::Peers::AutomationNotificationKind::ActionCompleted,
+                    Automation::Peers::AutomationNotificationProcessing::MostRecent,
+                    tb.Text(),
+                    L"KeyChordListenerText");
+            }
+        }
+    }
+
+    void KeyChordListener::KeyChordTextBox_KeyDown(IInspectable const& /*sender*/, KeyRoutedEventArgs const& e)
+    {
+        const auto key{ e.OriginalKey() };
+        for (const auto mod : ModifierKeys)
+        {
+            if (key == mod)
+            {
+                // Ignore modifier keys
+                return;
+            }
+        }
+
+        const auto modifiers{ _GetModifiers() };
```

Comment:
I don't understand the need for `_GetModifiers`.
If you keep track over how many keys are simultaneously pressed you can use the key down events directly to assemble the modifiers bitfield. This removes the need for the majority of the `_GetModifiers` implementation.
(You'd also need to reset the modifiers when the control looses focus of course.)

Reply:
My gut feeling is that introducing another place where state must be preserved and maintained is going to be worse for code clarity (and has the risk of introducing torn state) where here in a non-perf-sensitive context we can just use the state the UI framework is already keeping track of for us :smile:

Reply:
I haven't tried out this branch locally but I assumed that it works identical to key bindings in Discord: All key presses are additive, until you let go of any key, after which the result is added.

Meaning if you press:
* M Down
* Ctrl Down
* Shift Down
* M Up

...in that order it'll save Ctrl+Shift+M as a keybinding. Even though I pressed M first, which is important because the UI should work even if I attempt to press all 3 keys simultaneously! (...and then any of those may be pressed first.)

If that's how this is supposed to work as well I disagree: Using the events lends itself to a clearer association between what happens in the UI and how the code would work.

Reply:
I've tested it some more in Discord: If you let go of a modifier key for a while it'll count as a keybinding automatically. But I don't think we have to make it that complicated.

Also I just realized a problem with my approach: You still would have to detect when you have released all modifier keys. 😅 As such a good approach might be to count the concurrently pressed keys after all.
I'll approve the PR as I realize that the current code is already written and functional.

Reply:
I've been talking with Dustin about this: Making all key presses purely additive would probably be the most consistent way to enter key chords. Discord's keybinding editor is a pretty great template for us.
As I said before we'd need a "concurrently pressed keys counter" though, in order to detect when all keys have been released. Because every time this counter reaches zero we'll have to call `KeyChordSerialization::ToString` and see if all added up keys formed a valid sequence, as well as to reset the modifiers and vkey member fields.

Reply:
It's funny that you say that, because PowerToys' hotkey listener kinda works like that. It constantly updates the control on every key press, but it only sets the setting on a key-up event. 

I figured the current approach would be a bit easier and cleaner because we only accept key chords that have an associated non-modifier vkey. This also forces you to actually press the key combination you want to enact, because in TermControl pressing M-Ctrl-Shift would execute the M key binding.